essentials/src/api/dynamic.rs

```diff
@@ -0,0 +1,92 @@
+use super::subxt_wrapper::SubxtWrapperError::{self, DecodeDynamicError};
+use crate::metadata::{
+	polkadot::runtime_types::{
+		polkadot_parachain::primitives::Id,
+		polkadot_runtime_parachains::scheduler::{AssignmentKind, CoreAssignment},
+	},
+	polkadot_primitives::{CoreIndex, CoreOccupied, GroupIndex, ValidatorIndex},
+};
+use subxt::{
+	dynamic::{At, Value},
+	ext::scale_value::{Composite, Primitive, ValueDef, Variant},
+};
+
+pub(crate) fn decode_dynamic_validator_groups(
+	raw_groups: &Value<u32>,
+) -> Result<Vec<Vec<ValidatorIndex>>, SubxtWrapperError> {
+	let decoded_groups = decode_vector(raw_groups)?;
+	let mut groups = vec![];
+	for raw_group in decoded_groups.iter() {
+		let decoded_group = decode_vector(raw_group)?;
+		let mut group = vec![];
+		for raw_index in decoded_group.iter() {
+			group.push(ValidatorIndex(decode_u128_value(raw_index)? as u32));
+		}
+		groups.push(group)
+	}
+
+	Ok(groups)
+}
+
+pub(crate) fn decode_dynamic_availability_cores(
+	raw_cores: &Value<u32>,
+) -> Result<Vec<Option<CoreOccupied>>, SubxtWrapperError> {
+	let decoded_cores = decode_vector(raw_cores)?;
+	let mut cores = vec![];
+	for raw_core in decoded_cores.iter() {
+		cores.push(decode_option(raw_core)?.map(|v| match decode_variant(v).unwrap().name.as_str() {
+			"Parachain" => CoreOccupied::Parachain,
+			_ => todo!("Add parathreads support"),
+		}));
+	}
+
+	Ok(cores)
+}
+
+pub(crate) fn decode_dynamic_scheduled_paras(raw_paras: &Value<u32>) -> Result<Vec<CoreAssignment>, SubxtWrapperError> {
+	let decoded_paras = decode_vector(raw_paras)?;
+	let mut paras = vec![];
+	for para in decoded_paras.iter() {
```

Comment:
```suggestion
	for para in &decoded_paras {
```

Reply:
It doesn't work, because `decaded_paras` are `&Vec`, not `Vec`.